cmake/ccache.cmake

```diff
@@ -0,0 +1,93 @@
+WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+]]
+
+include_guard(GLOBAL)
+
+
+#[=======================================================================[.rst:
+enable_ccacche
+------------------
+
+Overview
+^^^^^^^^
+
+Enable use of Ccache for compiler caching during the build process.
+
+.. code-block:: cmake
+
+  enable_ccacche(
+      [QUIET]
+  )
+   -- Searches for the ccache executable, and if found enables it for supported
+   language with specified options.
+
+  ``QUIET`` Silent output from this method.
+
+#]=======================================================================]
+function(enable_ccache)
+    set(options QUIET)
+    set(oneValueArgs)
+    set(multiValueArgs)
+    cmake_parse_arguments(CCACHE "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN} )
+
+    find_program(CCACHE_BIN ccache REQUIRED)
+    if(NOT CCACHE_BIN)
+        message(STATUS "Morpheus: CCache not found. Rebuilding all the source files.")
+        message(DEBUG "Morpheus: If CCache compiler cache is desired then ensure this is a supported platform and ensure the Conan buildenv is active")
+        return()
+    endif()
+
+    message(STATUS "Morpheus Generator: ${CMAKE_GENERATOR}") #debug
+    message(STATUS "Morpheus: CCache found: ${CCACHE_BIN}. Enabling ccache as compiler cache tool.")
+    if(${CMAKE_GENERATOR} MATCHES "Ninja|Makefiles")
+        message(STATUS "found generator Ninja|Makefiles") #debug
+        foreach(lang IN ITEMS ${MORPHEUS_LANGUAGES})
+            set(CMAKE_${lang}_COMPILER_LAUNCHER ${CCACHE_BIN})
+        endforeach()
+    elseif(${CMAKE_GENERATOR} STREQUAL "Xcode")
+        message(STATUS "found generator Xcode") #debug
+        foreach(lang IN ITEMS ${MORPHEUS_LANGUAGES})
+            set(launch${lang} ${CMAKE_BINARY_DIR}/launch-${lang})
+            file(WRITE ${launch${lang}}
+                "#!/bin/bash\n\n"
+                "exec \"${CCACHE_BIN}\" \"${CMAKE_${lang}_COMPILER}\" \"$@\"\n"
+            )
+            execute_process(COMMAND chmod a+rx ${launch${lang}})
+            set(${CMAKE_XCODE_ATTRIBUTE_${lang}} ${launch${lang}})
+            if(lang STREQUAL "C")
+                set(CMAKE_XCODE_ATTRIBUTE_LD ${launchC}) 
+            elseif(lang STREQUAL "CXX")
+                set(CMAKE_XCODE_ATTRIBUTE_LDPLUSPLUS ${launchCXX})
+            endif()
+        endforeach()
+    elseif(${CMAKE_GENERATOR} MATCHES "Visual Studio")
+        message(STATUS "found generator Visual Studio") #debug
+        cmake_path(NATIVE_PATH CCACHE_BIN ccache_exe)
+        file(WRITE ${CMAKE_BINARY_DIR}/launch-cl.cmd
+        "@echo off\n"
+        "\"${ccache_exe}\" \"${CMAKE_C_COMPILER}\" %*\n"
```

Comment:
So it looks like we only generate a cmd invoke wrapper for the C compiler, is this correct?  I'd presume we need this for every language supported (I may be wrong here but be good to understand the rational)?

Reply:
Quoting chapter 35.4.4. Visual Studio Generators:
```
Using Ccache with the Visual Studio generators requires Ccache 4.6.1 or later. The
CMAKE_<LANG>_COMPILER_LAUNCHER variables don’t support Visual Studio generators, but Visual Studio
project properties can be used in conjunction with a launch script to achieve the same thing. The
CLToolPath and CLToolExe project properties redirect the compile steps to use a custom compiler
executable or script. The properties can be set with the CMAKE_VS_GLOBALS variable (see Section 35.3.3,
“Visual Studio Generators” for another closely related use of this variable).
The same compiler executable is used for C and C++ with these generators. Thus, only one launch
script is needed, unlike the Xcode generator which requires one launch script per language. 
```
That's the reason why I did it in this way. I could easily be wrong because I am a newbie and I am also developing on a Linux machine :smile:

Reply:
Ha, it may the the correct thing to do but I think it requires a little investigation and testing for each supported language.  I can tests on my Windows machine.